rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1562 @@
+                adr_name_gen::DeviceUpdateEventTelemetryReceiver::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &telemetry_options,
+                );
+            let asset_update_telemetry_receiver =
+                adr_name_gen::AssetUpdateEventTelemetryReceiver::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &telemetry_options,
+                );
+
+            async move {
+                Self::receive_update_notification_loop(
+                    shutdown_notifier_clone,
+                    device_update_telemetry_receiver,
+                    device_update_notification_dispatcher_clone,
+                    asset_update_telemetry_receiver,
+                    asset_update_notification_dispatcher_clone,
+                )
+                .await;
+            }
+        });
+
+        Ok(Self {
+            shutdown_notifier,
+            get_device_command_invoker: Arc::new(adr_name_gen::GetDeviceCommandInvoker::new(
+                application_context.clone(),
+                client.clone(),
+                &command_options,
+            )),
+            update_device_status_command_invoker: Arc::new(
+                adr_name_gen::UpdateDeviceStatusCommandInvoker::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &command_options,
+                ),
+            ),
+            notify_on_device_update_command_invoker: Arc::new(
+                adr_name_gen::SetNotificationPreferenceForDeviceUpdatesCommandInvoker::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &command_options,
+                ),
+            ),
+            device_update_notification_dispatcher,
+            get_asset_command_invoker: Arc::new(adr_name_gen::GetAssetCommandInvoker::new(
+                application_context.clone(),
+                client.clone(),
+                &command_options,
+            )),
+            update_asset_status_command_invoker: Arc::new(
+                adr_name_gen::UpdateAssetStatusCommandInvoker::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &command_options,
+                ),
+            ),
+            notify_on_asset_update_command_invoker: Arc::new(
+                adr_name_gen::SetNotificationPreferenceForAssetUpdatesCommandInvoker::new(
+                    application_context,
+                    client,
+                    &command_options,
+                ),
+            ),
+            asset_update_notification_dispatcher,
+        })
+    }
+
+    /// Convenience function to get all observed device & inbound endpoint names to quickly unobserve all of them before cleaning up
+    #[must_use]
+    pub fn get_all_observed_device_endpoints(&self) -> Vec<(String, String)> {
```

Comment:
are there any unit tests that can be added for these fns without network connectivity?